src/nano_contracts/deserializer.ts

```diff
@@ -192,53 +257,85 @@ class Deserializer {
-    const valueType = match ? match[1] : null;
-    if (!valueType) {
+  toSigned(signedData: Buffer, type: string): DeserializeResult<string> {
+    const [containerType, internalType] = getContainerInternalType(type);
+    if (containerType !== 'SignedData') {
+      throw new Error('Type is not SignedData');
+    }
+    if (!internalType) {
       throw new Error('Unable to extract type');
     }
+    // Should we check that the valueType is valid?
+
+    // SignData[T] is serialized as Serialize(T)+Serialize(sign(T)) where sign() returns a byte str
+    // Which means we can parse the T argument, then read the bytes after.
 
-    let signedBuffer: Buffer;
-    let size: number;
-    // [len(serializedResult)][serializedResult][inputData]
-    // eslint-disable-next-line prefer-const -- To split this declaration would be confusing
-    [size, signedBuffer] = unpackToInt(2, false, signedData);
-    let parsed = this.deserializeFromType(signedBuffer.slice(0, size), valueType);
-    if (valueType === 'bytes') {
+    // Reading argument
+    const parseResult = this.deserializeFromType(signedData, internalType);
+    let parsed = parseResult.value;
+    const bytesReadFromValue = parseResult.bytesRead;
+
+    if (internalType === 'bytes') {
       // If the value is bytes, we should transform into hex to return the string
-      parsed = (parsed as Buffer).toString('hex');
+      parsed = bufferToHex(parsed as Buffer);
+    }
+
+    if (internalType === 'bool') {
+      parsed = (parsed as boolean) ? 'true' : 'false';
     }
-    signedBuffer = signedBuffer.slice(size);
-    return `${bufferToHex(signedBuffer)},${parsed},${valueType}`;
+
+    // Reading signature
+    const { value: parsedSignature, bytesRead: bytesReadFromSignature } = this.deserializeFromType(
+      signedData.subarray(bytesReadFromValue),
+      'bytes'
+    );
+
+    return {
+      value: `${bufferToHex(parsedSignature as Buffer)},${parsed},${internalType}`,
+      bytesRead: bytesReadFromValue + bytesReadFromSignature,
+    };
   }
 
   /**
    * Deserialize a value decoded in bytes to a base58 string
    *
-   * @param {value} Value to deserialize
+   * @param {Buffer} buf Value to deserialize
    *
    * @memberof Deserializer
    * @inner
    */
-  toAddress(value: Buffer): string {
+  toAddress(buf: Buffer): DeserializeResult<string> {
+    const lenReadResult = leb128Util.decodeUnsigned(buf, 1);
+    if (lenReadResult.value !== 25n) {
+      // Address should be exactly 25 bytes long
+      throw new Error('Address should be 25 bytes long');
+    }
     // First we get the 20 bytes of the address without the version byte and checksum
-    const addressBytes = value.slice(1, 21);
+    const addressBytes = buf.subarray(2, 22);
     const address = helpersUtils.encodeAddress(addressBytes, this.network);
+    address.validateAddress();
     const decoded = address.decode();
-    if (decoded[0] !== value[0]) {
+    if (decoded[0] !== buf[1]) {
```

Comment:
question: Why do we compare bytes at different indexes in the decoded and in the original Buffer? We should have a comment explaining this.

Also, the error message does not show the same index for the `buf` Buffer.